src/nidas/dynld/raf/UDPiPMSensor.cc

```diff
@@ -0,0 +1,218 @@
+              "dev", "not found");
+        // addrInfo string contains addr, numphases, procquery, port,
+        // eg 0,1,5,30101
+        _addrInfo[i] = p->getStringValue(0);
+        ILOG(("device info for addresses %d: %s", i, _addrInfo[i].c_str()));
+    }
+
+}
+
+void UDPiPMSensor::open(int flags)
+{
+    UDPSocketSensor::open(flags);
+
+    _ctrl_pid = fork();
+
+    if (_ctrl_pid == -1)
+    {
+        PLOG(("UDPiPMSensor: error forking errno = %d", errno));
+    }
+    else
+    if (_ctrl_pid == 0)
+    {
+        char *args[40];
+        //char port[32];
+        char m_rate[32];
+        char r_period[32];
+        char baud[32];
+        char num_addr[32];
+        std::vector<std::string> addrnum(8);
+        int argc = 0;
+        char cmd[256] = "";
+        int cmd_len = 0;
+
+        args[argc++] = (char *)"ipm_ctrl";
+        if (_deviceAddr.length() > 0) {
+            args[argc++] = (char *)"-p";
+            args[argc++] = (char *)_deviceAddr.c_str();
+        }
+
+        /*if (_statusPort > 0) {
+            sprintf(port, "%u", _statusPort);
+            args[argc++] = (char *)"--port";
+            args[argc++] = (char *)port;
+        } */
+
+        if (_measureRate > 0) {
+            sprintf(m_rate, "%u", _measureRate);
+            args[argc++] = (char *)"-m";
+            args[argc++] = (char *)m_rate;
+        }
+
+        if (_recordPeriod > 0) {
+            sprintf(r_period, "%u", _recordPeriod);
+            args[argc++] = (char *)"-r";
+            args[argc++] = (char *)r_period;
+        }
+
+        if (_baudRate > 0) {
+            sprintf(baud, "%u", _baudRate);
+            args[argc++] = (char *)"-b";
+            args[argc++] = (char *)baud;
+        }
+
+        if (_numAddr > 0) {
+            sprintf(num_addr, "%u", _numAddr);
+            args[argc++] = (char *)"-n";
+            args[argc++] = (char *)num_addr;
+        }
+
+        for (int i=0; i < _numAddr; i++)
+        {
+            addrnum[i] = "-" + std::to_string(i);
+            args[argc++] = (char *)addrnum[i].c_str();
+            args[argc++] = (char *)_addrInfo[i].c_str();
+        }
+
+        args[argc] = (char *)0;
+
+        // Create a string containing the entire command for logging
+        // purposes.
+        for (int i=0; i < argc; i++)
+        {
+            cmd_len += strlen(args[i]) + 1;
+            if (cmd_len < 255)  // prevent memory overrun (unlikely)
+            {
+                strcat(cmd, args[i]);
+                strcat(cmd, " ");
+            }
+        }
+        ILOG(("UDPiPMSensor: forking command %s", cmd));
+
+        if (execvp(args[0], args) == -1)
+        {
+            ILOG(("UDPiPMSensor: error executing command %s", cmd));
+            exit(1);
+        } else {
+            ILOG(("UDPiPMSensor: success!!! %s", cmd));
+        }
+    }
+}
+
+void UDPiPMSensor::close()
+{
+    UDPSocketSensor::close();
+
+    if (_ctrl_pid > 0)
+    {
+        int rc = kill(_ctrl_pid, SIGTERM);
```

Comment:
I think this is confusing because the `kill()` return value is a status which is not checked, and then it is passed into `wait()` as the return value for a PID.  Also, `wait()` could return on any exited child, not just _ctrl_pid, so I assume what this is really trying to do is something like `waitpid(_ctrl_pid, 0, 0)`, or else don't bother waiting at all.

I realize this code has been copied from other places, so I point it out to try to keep it from propagating further... :)

Reply:
if (_ctrl_pid > 0)
{
    if (kill(_ctrl_pid, SIGTERM) == -1)
    {
        ELOG(("UDPArincSensor: kill() error: ") << ": error " << errno
                    << " : " << Exception::errnoToString(errno));
    }

    if (waitpid(_ctrl_pid, 0, 0) == -1)
    {   
        ELOG(("UDPArincSensor: waitpid() error: ") << ": error " << errno
                    << " : " << Exception::errnoToString(errno));
    }
}
Should do anything other than log failure? I assume we don't really want the dsm process to exit. And do we wait if the kill failed?

Reply:
Should do anything other than log failure? I assume we don't really want the dsm process to exit. And do we wait if the kill failed?

    if (_ctrl_pid > 0)
    {
        if (kill(_ctrl_pid, SIGTERM) == -1)
        {
            ELOG(("UDPArincSensor: kill() error: ") << ": error " << errno
                        << " : " << Exception::errnoToString(errno));
        }

        if (waitpid(_ctrl_pid, 0, 0) == -1)
        {   
            ELOG(("UDPArincSensor: waitpid() error: ") << ": error " << errno
                        << " : " << Exception::errnoToString(errno));
        }
    }

Reply:
Those are difficult questions.  If the kill fails then probably you have bigger problems than a hanging child, and you'll certainly want to know about the error.  Even if the kill fails, the child may have exited and could be collected with a `wait()`, so you could test for that with the `WNOHANG` option to `waitpid()`.  It's ok to wait for a process which has already exited, the wait will just immediately return, and apparently it's ok to send a signal a child process that has already exited, because a child which has not been waited for still exists as a zombie.  As to whether to wait at all, I assume it depends on how important it is that the child exits cleanly and whether the child's exit needs to be synchronized with the parent, ie, the parent needs to avoid doing something while the child process still exists, like starting another child.  Since this is in sensor `close()`, is it possible the `close()` is being called as part of a re-open?  It seems possible that the parent dsm process could accumulate a bunch of zombie children on each re-open if the parent does not `wait()` for them.  And if the child process is hanging or crashing rather than exiting cleanly on the signal, then that's probably something you want to notice sooner rather than later.  So that means always calling `wait()` and also logging non-zero exit statuses from the child.  That seems like a safe and robust option to me.